platforms/allwinner-d1/d1-core/src/drivers/smhc.rs

```diff
@@ -0,0 +1,874 @@
+                    self.state
+                }
+            }
+        };
+
+        if needs_wake {
+            if let Some(waker) = self.waker.take() {
+                waker.wake();
+                // If we are waking the driver task, we need to disable interrupts
+                smhc.smhc_ctrl.modify(|_, w| w.ine_enb().disable());
+            }
+        }
+    }
+}
+
+/// Internal DMA controller
+mod idmac {
+    use core::mem;
+    use core::ptr::NonNull;
+    use mycelium_bitfield::bitfield;
+
+    /// A descriptor that describes how memory needs to transfer data
+    /// between the SMHC port and host memory. Multiple DMA transfers
+    /// can be configured by creating a descriptor *chain* (linked list).
+    #[derive(Clone, Debug)]
+    #[repr(C, align(4))]
+    pub(super) struct Descriptor {
+        /// The descriptor configuration.
+        configuration: Cfg,
+        /// The size of the data buffer.
+        buff_size: BuffSize,
+        /// The (*word*) address of the data buffer.
+        buff_addr: u32,
+        /// The (*word*) address of the next descriptor, for creating a descriptor chain.
+        next_desc: u32,
+    }
+
+    /// A builder for constructing an IDMAC [`Descriptor`].
+    #[derive(Copy, Clone, Debug)]
+    #[must_use = "a `DescriptorBuilder` does nothing unless `DescriptorBuilder::build()` is called"]
+    pub(super) struct DescriptorBuilder<B = ()> {
+        cfg: Cfg,
+        buff: B,
+        link: u32,
+    }
+
+    #[derive(Debug)]
+    pub(super) enum Error {
+        BufferAddr,
+        BufferSize,
+        Link,
+    }
+
+    bitfield! {
+        /// The first 32-bit word of an IDMAC descriptor, containing configuration data.
+        struct Cfg<u32> {
+            // Bit 0 is reserved.
+            const _RESERVED_0 = 1;
+
+            /// Disable interrupts on completion.
+            const CUR_TXRX_OVER_INT_DIS = 1;
+
+            /// When set to 1, this bit indicates that the buffers this descriptor points to
+            /// are the last data buffer.
+            const LAST_FLAG = 1;
+
+            /// When set to 1, this bit indicates that this descriptor contains
+            /// the first buffer of data. It must be set to 1 in the first DES.
+            const FIRST_FLAG = 1;
+
+            /// When set to 1, this bit indicates that the second address in the descriptor
+            /// is the next descriptor address. It must be set to 1.
+            const CHAIN_MOD = 1;
+
+            /// Bits 29:5 are reserved.
+            const _RESERVED_1 = 25;
+
+            /// When some errors happen in transfer, this bit will be set to 1 by the IDMAC.
+            const ERR_FLAG = 1;
+
+            /// When set to 1, this bit indicates that the descriptor is owned by the IDMAC.
+            /// When this bit is reset, it indicates that the descriptor is owned by the host.
+            /// This bit is cleared when the transfer is over.
+            const DES_OWN_FLAG = 1;
```

Comment:
nit, take it or leave it: i would probably use `bool` for the various 1-bit flags here, rather than a single-bit numeric bitfield:
```suggestion
            const _RESERVED_0 = 1;

            /// Disable interrupts on completion.
            const CUR_TXRX_OVER_INT_DIS = 1;

            /// When set to 1, this bit indicates that the buffers this descriptor points to
            /// are the last data buffer.
            const LAST_FLAG: bool;

            /// When set to 1, this bit indicates that this descriptor contains
            /// the first buffer of data. It must be set to 1 in the first DES.
            const FIRST_FLAG: bool;

            /// When set to 1, this bit indicates that the second address in the descriptor
            /// is the next descriptor address. It must be set to 1.
            const CHAIN_MOD bool;

            /// Bits 29:5 are reserved.
            const _RESERVED_1 = 25;

            /// When some errors happen in transfer, this bit will be set to 1 by the IDMAC.
            const ERR_FLAG: bool;

            /// When set to 1, this bit indicates that the descriptor is owned by the IDMAC.
            /// When this bit is reset, it indicates that the descriptor is owned by the host.
            /// This bit is cleared when the transfer is over.
            const DES_OWN_FLAG:  bool;
```

stylistically, i would also probably drop `_FLAG` from these names, but that's up to you.